core/platform/email/gae_email_services_test.py

```diff
@@ -23,23 +23,310 @@
+        # Tests that email has no bcc attribute if bcc_admin is False.
+        messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html', bcc_admin=False)
+        messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
+        self.assertEqual(1, len(messages))
+        self.assertFalse(hasattr(messages[0], 'bcc'))
+
+    def test_sending_email_with_reply_to_id_adds_reply_to_email(self):
+        # Tests that email has reply_to_address if reply_to_id is passed.
+        messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+
+        reply_to_id = 'reply_to_id'
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html', reply_to_id=reply_to_id)
+        messages = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAIL)
+        expected_reply_to_address = 'reply+reply_to_id@example.com'
+        self.assertEqual(messages[0].reply_to, expected_reply_to_address)
+
+    def test_sending_email_without_reply_to_id_not_add_reply_to_email(self):
+        """Tests that email does not have reply_to_address if reply_to_id is
+        not passed.
+        """
+        messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True):
+            gae_email_services.send_mail(
+                self.SENDER_EMAIL, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html')
+        messages = self.mail_stub.get_sent_messages(
+            to=self.RECIPIENT_EMAIL)
+        self.assertFalse(hasattr(messages[0], 'reply_to'))
+
+    def test_email_not_sent_if_sender_address_is_malformed(self):
+        messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+        malformed_sender_email = ''
+        email_exception = self.assertRaisesRegexp(
+            ValueError, 'Malformed sender email address: %s'
+            % malformed_sender_email)
+        with self.swap(feconf, 'CAN_SEND_EMAILS', True), email_exception:
+            gae_email_services.send_mail(
+                malformed_sender_email, self.RECIPIENT_EMAIL,
+                'subject', 'body', 'html')
+        messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+
+        messages = self.mail_stub.get_sent_messages(to=self.RECIPIENT_EMAIL)
+        self.assertEqual(0, len(messages))
+        malformed_sender_email = 'malformed_sender_email@example.com'
```

Comment:
Here and elsewhere, give a hint in the test data about **why** it's malformed. E.g. call this missing_email_name@example.com and the following one MissingEmailAddress. There are probably others you can clarify.

Reply:
Sure. I will add comments explaining the reasons why these are malformed.
Thanks!

Reply:
Done!